rust/connlib/tunnel/src/peer.rs

```diff
@@ -356,30 +358,45 @@ impl ClientOnGateway {
     // in case that 2 or more resources have overlapping rules.
     fn recalculate_filters(&mut self) {
         self.filters = IpNetworkTable::new();
-        for resource in self.resources.values() {
+        self.recalculate_cidr_filters();
+        self.recalculate_dns_filters();
+
+        self.internet_resource_enabled = self.resources.values().any(|r| r.is_internet_resource());
+    }
+
+    fn recalculate_cidr_filters(&mut self) {
+        for resource in self.resources.values().filter(|r| r.is_cidr()) {
             for ip in &resource.ips() {
-                let mut filter_engine = FilterEngine::empty();
                 let filters = self.resources.values().filter_map(|r| {
                     r.ips()
                         .iter()
                         .any(|r_ip| network_contains_network(*r_ip, *ip))
                         .then_some(r.filters())
                 });
 
-                // Empty filters means permit all
-                if filters.clone().any(|f| f.is_empty()) {
-                    filter_engine.permit_all();
-                }
-
-                filter_engine.add_filters(filters.flatten());
+                let filter_engine = FilterEngine::with_filters(filters);
 
                 tracing::trace!(%ip, filters = ?filter_engine, "Installing new filters");
-
                 self.filters.insert(*ip, filter_engine);
             }
         }
     }
 
+    fn recalculate_dns_filters(&mut self) {
+        for (addr, TranslationState { resource_id, .. }) in &self.permanent_translations {
+            let Some(resource) = self.resources.get(resource_id) else {
+                continue;
+            };
+
+            debug_assert!(resource.is_dns());
+
+            let filter_engine = FilterEngine::with_filters(iter::once(resource.filters()));
+
+            tracing::trace!(%addr, filters = ?filter_engine, "Installing new filters");
+            self.filters.insert(*addr, filter_engine);
```

Comment:
Could we deduplicate this too? I think it should be possible to have have a function that takes ip and list of filters, creates a filter engine, logs it and inserts it, right?

Reply:
It's possible but I'd need a wrapper for the network table so it can be done by `self.filter` or a static function. Otherwise the borrow checker won't allow to take `self` as mutable as I'm already holding `self.resources` as immutable.

I could also just clone `self.resources` but would require making `ResourceOnGateway` clonable.

Reply:
updated with the static function version